wasm/src/lsp/types.rs

```diff
@@ -352,3 +355,138 @@ fn test_doc_range_overlap_at_zero() {
         end: DocPosition { line: 0, character: 3 }
     }));
 }
+
+// An object that contains the literal text of a document we're working with in
+// the LSP.
+#[derive(Debug, Clone)]
+pub struct DocData {
+    pub fullname: String,
+    pub text: Vec<Rc<Vec<u8>>>,
+    pub version: i32,
+    // Zero based.
+    pub comments: HashMap<usize, usize>,
+}
+
+impl DocData {
+    // Return a reference to the nth line's data.
+    pub fn nth_line_ref(&self, line: usize) -> Option<&Vec<u8>> {
+        if line < self.text.len() {
+            let borrowed: &Vec<u8> = self.text[line].borrow();
+            Some(borrowed)
```

Comment:
(Disclaimer: I'm not a pro Rustacian)

`RC<T>` seems to already implements `borrow::Borrow` so 
```rust
let borrowed: &Vec<u8> = self.text[line].borrow();
Some(borrowed)
````
Can be simplified to
```rust
Some(&self.text[line])
```
and `use` statement below can be removed.
```rust
use std::borrow::Borrow;
```

![image](https://user-images.githubusercontent.com/84098616/209343310-d4f54c19-5876-4412-b857-52854c7c5980.png)

Reply:
I almost always do it this way because there is an intermediate type that rust can't always guess.
If the result object is passed into a trait method then the type is often ambiguous.  I think of this as a style choice that reduces cognitive load (for me anyway), but if it seems like a big deal i can comb through this code and find all the places where ambiguity doesn't appear and change them if really desired.